### PR TITLE
Constify API, a bit more ...

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
Pretty straightforward, Travis was ok.
I can reduce its extent, if you think its too much or too late.

### CODE REVIEW

File: crypto/x509v3/v3_utl.c [hunk 1]
```diff
@@ -840,8 +840,8 @@ static int do_x509_check(X509 *x, const char *chk, size_t chklen,
     i = -1;
     name = X509_get_subject_name(x);
     while ((i = X509_NAME_get_index_by_NID(name, cnid, i)) >= 0) {
-        X509_NAME_ENTRY *ne;
-        ASN1_STRING *str;
+        const X509_NAME_ENTRY *ne;
+        const ASN1_STRING *str;
         ne = X509_NAME_get_entry(name, i);
```
Comment (richsalz) [inline crypto/x509v3/v3_utl.c]: merge the assignment and declaration and add a blank line.

File: crypto/x509/x509name.c [hunk 2]
```diff
@@ -214,7 +214,11 @@ int X509_NAME_add_entry(X509_NAME *name, X509_NAME_ENTRY *ne, int loc,
         inc = (set == 0) ? 1 : 0;
     }
 
-    if ((new_name = X509_NAME_ENTRY_dup(ne)) == NULL)
+    /*
+     * X509_NAME_ENTRY_dup is ASN1 generated code, that can't be easily 
+     * const'ified ; harmless cast as dup() don't modify its input.
+     */
+    if ((new_name = X509_NAME_ENTRY_dup((X509_NAME_ENTRY *)ne)) == NULL)
```
Comment (richsalz) [inline crypto/x509/x509name.c]: nice comment.  and the 'ugl...but harmless' comments elsewhere.

Comment (FdaSilvaYY) [inline crypto/x509/x509name.c]: Fixed ;)

File: crypto/asn1/a_dup.c [hunk 3]
```diff
@@ -56,7 +56,7 @@ void *ASN1_item_dup(const ASN1_ITEM *it, void *x)
     if (x == NULL)
         return (NULL);
 
-    i = ASN1_item_i2d(x, &b, it);
+    i = ASN1_item_i2d((ASN1_VALUE *)x, &b, it);
```
Comment (mattcaswell) [inline crypto/asn1/a_dup.c]: Why is this necessary?

File: crypto/asn1/a_strex.c [hunk 4]
```diff
@@ -454,10 +454,14 @@ static int do_name_ex(char_io *io_ch, void *arg, X509_NAME *n,
 
     cnt = X509_NAME_entry_count(n);
     for (i = 0; i < cnt; i++) {
+        /*
+         * Ugly casts to (X509_NAME *) ... but harmless 
+         * Waiting for the const'ification of STACK_OF api
+         */
         if (flags & XN_FLAG_DN_REV)
-            ent = X509_NAME_get_entry(n, cnt - i - 1);
+            ent = X509_NAME_get_entry((X509_NAME *)n, cnt - i - 1);
```
Comment (mattcaswell) [inline crypto/asn1/a_strex.c]: It seems straight forward to change `X509_NAME_get_entry` to accept a `const`?

**Comment by mattcaswell** [comment]
Just a couple of minor comments. Otherwise looks good.

File: crypto/asn1/a_strex.c (hunk 4 above)
Comment (FdaSilvaYY) [inline crypto/asn1/a_strex.c]: Unless there is 20 calling point ;)
I'm doing some test about it of a separated branch, waiting for Travis reply.

Comment (FdaSilvaYY) [inline crypto/asn1/a_strex.c]: Ping @mattcaswell : fixed in a new separated commit.
As stated, it is quite large ;)

File: crypto/x509/x509name.c [hunk 5]
```diff
@@ -334,14 +338,14 @@ int X509_NAME_ENTRY_set_data(X509_NAME_ENTRY *ne, int type,
     return (1);
 }
 
-ASN1_OBJECT *X509_NAME_ENTRY_get_object(X509_NAME_ENTRY *ne)
+const ASN1_OBJECT *X509_NAME_ENTRY_get0_object(const X509_NAME_ENTRY *ne)
```
Comment (mattcaswell) [inline crypto/x509/x509name.c]: Sorry, I must have missed this before. We can't make this change. `X509_NAME_ENTRY_get_object` exists in 1.0.2, so we can't just delete it. We can add the `X509_NAME_ENTRY_get0_object` *as well* if we choose to.

File: crypto/x509/x509name.c [hunk 6]
```diff
@@ -334,14 +338,14 @@ int X509_NAME_ENTRY_set_data(X509_NAME_ENTRY *ne, int type,
...
 {
     if (ne == NULL)
         return (NULL);
     return (ne->object);
 }
 
-ASN1_STRING *X509_NAME_ENTRY_get_data(X509_NAME_ENTRY *ne)
+const ASN1_STRING *X509_NAME_ENTRY_get0_data(const X509_NAME_ENTRY *ne)
```
Comment (mattcaswell) [inline crypto/x509/x509name.c]: Same here. This function exists in 1.0.2

File: apps/ca.c [hunk 7]
```diff
@@ -1412,44 +1411,45 @@ static int do_body(X509 **xret, EVP_PKEY *pkey, X509 *x509,
...
     for (i = 0; i < X509_NAME_entry_count(name); i++) {
-        ne = X509_NAME_get_entry(name, i);
-        str = X509_NAME_ENTRY_get_data(ne);
-        obj = X509_NAME_ENTRY_get_object(ne);
+        ne = X509_NAME_get0_entry(name, i);
+        str = X509_NAME_ENTRY_get0_data(ne);
+        obj = X509_NAME_ENTRY_get0_object(ne);
 
         if (msie_hack) {
+            /* Dirty ugly cast to make the msie_hack compile!! */
```
Comment (mattcaswell) [inline apps/ca.c]: I don't like this at all. We need to retain a getter for a  mutable `ASN1_STRING` until we can sort this out.

File: apps/ca.c [hunk 8]
```diff
@@ -1412,44 +1411,45 @@ static int do_body(X509 **xret, EVP_PKEY *pkey, X509 *x509,
...
-            && (str->type != V_ASN1_UTF8STRING)) {
+        if (str->type != V_ASN1_BMPSTRING && str->type != V_ASN1_UTF8STRING) {
             j = ASN1_PRINTABLE_type(str->data, str->length);
-            if (((j == V_ASN1_T61STRING) &&
-                 (str->type != V_ASN1_T61STRING)) ||
-                ((j == V_ASN1_IA5STRING) &&
-                 (str->type == V_ASN1_PRINTABLESTRING))) {
+            if ((j == V_ASN1_T61STRING && str->type != V_ASN1_T61STRING)
+                ||
+                (j == V_ASN1_IA5STRING && str->type == V_ASN1_PRINTABLESTRING)) {
```
Comment (mattcaswell) [inline apps/ca.c]: This looks very odd.

File: crypto/x509/x509name.c (hunk 5 above)
Comment (mattcaswell) [inline crypto/x509/x509name.c]: Oh...I see why I didn't notice this before -  this is the new commit you introduced. If this was in response to my previous comment, all I really expected was this:
````
ASN1_OBJECT *X509_NAME_ENTRY_get_object(const X509_NAME_ENTRY *ne);
````

File: apps/ca.c [hunk 9]
```diff
@@ -1558,7 +1557,7 @@ static int do_body(X509 **xret, EVP_PKEY *pkey, X509 *x509,
 
             if (push != NULL) {
                 if (!X509_NAME_add_entry(subject, push, -1, 0)) {
-                    X509_NAME_ENTRY_free(push);
+                    /* useless X509_NAME_ENTRY_free(push); */
```
Comment (mattcaswell) [inline apps/ca.c]: Why comment this out? The original code just looks wrong AFAICT - shouldn't we just delete it?

File: crypto/asn1/a_strex.c [hunk 10]
```diff
@@ -459,9 +459,9 @@ static int do_name_ex(char_io *io_ch, void *arg, const X509_NAME *n,
          * Waiting for the const'ification of STACK_OF api
          */
```
Comment (mattcaswell) [inline crypto/asn1/a_strex.c]: The ugly casts comment can go now

File: crypto/x509/x509name.c [hunk 11]
```diff
@@ -82,7 +82,7 @@ int X509_NAME_get_index_by_OBJ(X509_NAME *name, const ASN1_OBJECT *obj, int last
     return (-1);
 }
 
-X509_NAME_ENTRY *X509_NAME_get_entry(X509_NAME *name, int loc)
+const X509_NAME_ENTRY *X509_NAME_get0_entry(const X509_NAME *name, int loc)
```
Comment (mattcaswell) [inline crypto/x509/x509name.c]: This one also existed in 1.0.2

**Comment by mattcaswell** [comment]
I've stopped reviewing the new commit at this point. I think as far as this MR is concerned you should make the minimal fix I proposed above:
````
ASN1_OBJECT *X509_NAME_ENTRY_get_object(const X509_NAME_ENTRY *ne);
````
The new commit should be raised as a separate MR as it is so big.

File: crypto/asn1/a_strex.c [hunk 12]
```diff
@@ -454,6 +454,10 @@ static int do_name_ex(char_io *io_ch, void *arg, X509_NAME *n,
 
     cnt = X509_NAME_entry_count(n);
     for (i = 0; i < cnt; i++) {
+        /*
+         * Ugly casts to (X509_NAME *) ... but harmless 
+         * Waiting for the const'ification of STACK_OF api
+         */
```
Comment (mattcaswell) [inline crypto/asn1/a_strex.c]: This comment is no longer relevant.

**Comment by mattcaswell** [comment]
+1 subject to the out-of-date comment noted above being removed.

File: crypto/x509v3/v3_ncons.c [hunk 13]
```diff
@@ -199,7 +199,7 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
         /* Process any email address attributes in subject name */
 
         for (i = -1;;) {
-            X509_NAME_ENTRY *ne;
+            const X509_NAME_ENTRY *ne;
```
Comment (vdukhovni) [inline crypto/x509v3/v3_ncons.c]: Please add a blank line after the declarations.

**Comment by vdukhovni** [comment]
Apart from missing blank line (which was the case even before this PR, but should be addressed) looks fine to me.

**Comment by FdaSilvaYY** [comment]
@mattcaswell : fixed and rebase. please proceed.

**Comment by levitte** [comment]
+1

**Comment by levitte** [comment]
Merged
